openedx_learning/apps/authoring/publishing/api.py

```diff
@@ -457,13 +490,214 @@ def set_draft_version(
+                new_version_id=publishable_entity_version_pk,
+            )
+            _create_container_side_effects_for_draft_change(change)
+
+
+@set_draft_version.register(int)
+def _(
+    publishable_entity_id: int,
+    publishable_entity_version_pk: int | None,
+    /,
+    set_at: datetime | None = None,
+    set_by: int | None = None,  # User.id
+    create_transaction: bool = True,
+) -> None:
+    """
+    Alias for set_draft_version taking PublishableEntity.id instead of a Draft.
+    """
+    tx_context = atomic() if create_transaction else nullcontext()
+    with tx_context:
+        draft, _created = Draft.objects.select_related("entity") \
+                                        .get_or_create(entity_id=publishable_entity_id)
+        set_draft_version(
+            draft,
+            publishable_entity_version_pk,
+            set_at=set_at,
+            set_by=set_by,
+            create_transaction=False,
+        )
+
+
+def _add_to_existing_draft_change_log(
+    active_change_log: DraftChangeLog,
+    entity_id: int,
+    old_version_id: int | None,
+    new_version_id: int | None,
+) -> DraftChangeLogRecord:
+    """
+    Create or update a DraftChangeLogRecord for the active_change_log passed in.
+
+    The an active_change_log may have many DraftChangeLogRecords already
+    associated with it. A DraftChangeLog can only have one DraftChangeLogRecord
+    per PublishableEntity, e.g. the same Component can't go from v1 to v2 and v2
+    to v3 in the same DraftChangeLog. The DraftChangeLogRecord is meant to
+    capture the before and after states of the Draft version for that entity,
+    so we always keep the first value for old_version, while updating to the
+    most recent value for new_version.
+
+    So for example, if we called this function with the same active_change_log
+    and the same entity_id but with versions: (None, v1), (v1, v2), (v2, v3);
+    we would collapse them into one DraftChangeLogrecord with old_version = None
+    and new_version = v3.
+    """
+    try:
+        # If there is already a DraftChangeLogRecord for this PublishableEntity,
+        # we update the new_version_id. We keep the old_version_id value
+        # though, because that represents what it was before this
+        # DraftChangeLog, and we don't want to lose that information.
+        change = DraftChangeLogRecord.objects.get(
+            draft_change_log=active_change_log,
+            entity_id=entity_id,
+        )
+        change.new_version_id = new_version_id
+        change.save()
+    except DraftChangeLogRecord.DoesNotExist:
+        # If we're here, this is the first DraftChangeLogRecord we're making for
+        # this PublishableEntity in the active DraftChangeLog.
+        change = DraftChangeLogRecord.objects.create(
+            draft_change_log=active_change_log,
+            entity_id=entity_id,
+            old_version_id=old_version_id,
+            new_version_id=new_version_id,
+        )
```

Comment:
Nit: could use `update_or_create` here to simplify this a bit.

Reply:
I don't think it quite works because `create_defaults` isn't supported as an argument in Django 4.2. We want to set both `old_version` and `new_version` if we're creating, but only touch `new_version` if we're updating.

Reply:
Ah, OK. Didn't check the older docs.